### PR TITLE
Configurable cryptography

### PR DESCRIPTION
**Pull request by thomcc**
By default it has the `use_ring` feature enabled, and uses ring. Someone who wants to use openssl can turn on the `use_openssl` feature, and someone who wants neither (e.g. mozilla/application-services) can turn off both and provide a custom impl with set_cryptographer.

Since there's nothing sane we can do if both features are on, we detect it and give a explicit error message in a build.rs.

Annoyingly, this means testing / linting both requires

```
cargo test
cargo test --no-default-features --features 'use_openssl'
```

Cargo doesn't really support mutually exclusive features well.

Note: this needs the latest stable version of Rust to compile (e.g. the one that came out yesterday). I don't know if anything needs to change in the tc config for that to work.

Supports https://github.com/mozilla/application-services/issues/959, assuming @eoger thinks we can shoehorn what we need into the trait I've defined there.

There's probably a way this could have less overhead in the ring/openssl cases, but it would be harder to test, so I didn't go with that. As it is, the dynamic machinery is used every time.

### CODE REVIEW

**Comment by thomcc** [comment]
Note: let's not land this before @eoger gets a chance to review.

**Comment by djmitche** [comment]
Sure thing.  I'd like to get the CI updated before we land, too.

**Comment by thomcc** [comment]
Given the account that seems prudent :P.

I'll probably need a pointer on how to do that though, unless slapping in a `rustup update` is acceptable.

File: src/crypto/holder.rs
Comment (djmitche) [inline src/crypto/holder.rs:9]: I would think this could just be a panic?  It's definitely a programming error if this occurs, right? It looks like the idea is that the error can be ignored in `autoinit_crypto`, but that seems wrong, too -- if the crate is built with one of the `use_` features, then there should be a panic when the `autoinit_crypto` for that feature is called after `set_cryptographer` is called.  In fact, in the case that either feature is set it probably makes sense to hide the `set_cryptographer` functions.

File: src/error.rs
Comment (djmitche) [inline src/error.rs:30]: This might warrant mention in the upgrade notes, too?

File: src/payload.rs
Comment (djmitche) [inline src/payload.rs:11]: Another compatibility note.

Alternately, it looks like `CryptoError::UnsupportedDigest` is the only error that might be returned here, and that's arguably a programming error in this library or the custom `Cryptographer` implementation and thus suitable for a panic.  But I suppose at some point some Cryptographer implementation could conceivably want to signal an `Err`, so it probably makes sense to add `Result` now.

File: src/payload.rs
Comment (djmitche) [inline src/payload.rs:27]: Compatibility note

File: src/credentials.rs
Comment (djmitche) [inline src/credentials.rs:24]: Compatibility note

File: src/credentials.rs
Comment (djmitche) [inline src/credentials.rs:31]: Compatibility note

**Review by djmitche — COMMENTED** [review]
Let's add a section to the README about this, perhaps "Upgrading", that notes the changes here (seems mostly just things now returning `Result`, but also `Key::new` no longer taking a reference).

I guess the "dynamic machinery" you indicate is `dyn HmacKey` etc., meaning that crypto operations involve a vtable call or two.  That's sort of sad since it's in the hot path for this library, but it's a hot path that involves a bunch of crypto math so an indirect jump or two probably isn't a big deal.

Otherwise, I'm quite happy with this.  I'll fix up the CI in another push, but don't let that slow you down.

File: src/crypto/holder.rs
Comment (thomcc) [reply src/crypto/holder.rs:9]: I more or less based this bit on the API for setting the global logger in the `log` crate.

More generally, while it is a programming error:

- There's no API exposed to determine if this function has been called, and I could imagine situations where it's tricky for to be certain about it.
- I'd also prefer to push panics into the consumers, so it's obvious to them that this panics on failure.

File: src/error.rs
Comment (thomcc) [reply src/error.rs:30]: Yes, this is a backwards incompatible change. I'll write up a set of changes on Monday.

File: src/payload.rs
Comment (thomcc) [reply src/payload.rs:11]: Yeah, I don't think this is suitable for a panic TBH. It's reasonable to expect that all cryptographers don't support all algos here, and since there's no way to detect if the cryptographer supports a specific algo, yeah.

File: CHANGELOG.md
Comment (djmitche) [inline CHANGELOG.md:31]: :+1:

**Review by djmitche — APPROVED** [review]

File: src/crypto/holder.rs
Comment (eoger) [inline src/crypto/holder.rs:19]: Maybe add a comment why this is OK to leak the `Box` as I'm just trusting you know your rust-foo enough to convince me this is fine.

**Review by eoger — APPROVED** [review]
Crypto parts looks good to me, thanks a lot for taking on that work Thom

**Comment by djmitche** [comment]
I'll get this merged, but I'd like #27 to get merged too before making a release, since otherwise that PR would constitute a further breaking change.

**Comment by rfk** [comment]
Hi @djmitche, anything else we can do to help with a new release to include this change?

**Comment by djmitche** [comment]
I think we can do that.  I have another PR in the works to support signing `http.request.Builder` values, that is another breaking change, but there are after all a countably infinite set of possible values for the X in X.Y.Z :)

**Comment by djmitche** [comment]
v3.0.0 should be on crates.io shortly!

**Pull request merged**
